pkg/source/firefly/source.go

```diff
@@ -0,0 +1,187 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package firefly
+
+import (
+	"context"
+	"strings"
+
+	"github.com/TheThingsNetwork/go-utils/random"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	"go.thethings.network/lorawan-stack-migrate/pkg/source"
+	"go.thethings.network/lorawan-stack-migrate/pkg/source/firefly/client"
+	"go.thethings.network/lorawan-stack-migrate/pkg/util"
+)
+
+type Source struct {
+	*Config
+	*client.Client
+}
+
+func createNewSource(cfg *Config) source.CreateSource {
+	return func(ctx context.Context, src source.Config) (source.Source, error) {
+		cfg.src = src
+		if err := cfg.Initialize(); err != nil {
+			return nil, err
+		}
+		client, err := cfg.NewClient(logger)
+		if err != nil {
+			return nil, err
+		}
+		return Source{
+			Config: cfg,
+			Client: client,
+		}, nil
+	}
+}
+
+// ExportDevice implements the source.Source interface.
+func (s Source) ExportDevice(devEUI string) (*ttnpb.EndDevice, error) {
+	ffdev, err := s.GetDeviceByEUI(devEUI)
+	if err != nil {
+		return nil, err
+	}
+	if ffdev == nil {
+		return nil, errNoDeviceFound.WithAttributes("eui", devEUI)
+	}
+	v3dev := &ttnpb.EndDevice{
+		Name:            ffdev.Name,
+		Description:     ffdev.Description,
+		FrequencyPlanId: s.frequencyPlanID,
```

Comment:
So I guess the reason why the uplink channel 867.7 MHz is not found, is that the uplink channels aren't set as the current parameters, right? The frequency plan alone is not enough.

Reply:
Correct - the end device is configured by Firefly to operate on 8 channels, but since these channels are not in the current parameters, matching fails.